### PR TITLE
Test the ambiguous ampersand state

### PR DESCRIPTION
**Pull request by zcorpan**
Follows https://github.com/whatwg/html/pull/2731

---

I couldn't figure out how to run the tokenizer tests with html5lib. Just `nosetests` doesn't seem to run them?

### CODE REVIEW

**Comment by inikulin** [comment]
I believe it makes to use new error format for this

**Comment by zcorpan** [comment]
I took `unknown-named-character-reference` from the spec and otherwise made it consistent with surrounding tests. I figured we may want to sweep all tests to update to the new format. But happy to make the new tests added here use the new format.

What should the error lines say, exactly?

**Comment by zcorpan** [comment]
Sorry, that's what https://github.com/html5lib/html5lib-tests/pull/92 does. So, I can rebase this on 92 I suppose.

**Comment by zcorpan** [comment]
Or vice versa; if this looks OK we can land this first and rebase 92 and convert these tests there. WDYT?

**Comment by inikulin** [comment]
@zcorpan Both works for me. Do whatever will take less efforts from you.

**Comment by zcorpan** [comment]
OK, I think I prefer landing this one first.

**Comment by gsnedders** [comment]
@zcorpan You need to update the submodule, make sure you have dependencies installed (`pip install -r requirements-test.txt`), and run `pytest`.

File: tokenizer/entities.test [hunk 1]
```diff
@@ -2,12 +2,24 @@
 
 {"description": "Undefined named entity in attribute value ending in semicolon and whose name starts with a known entity name.",
 "input":"<h a='&noti;'>",
-"output": [["StartTag", "h", {"a": "&noti;"}]]},
+"output": ["ParseError", ["StartTag", "h", {"a": "&noti;"}]]},
```
Comment (inikulin) [inline tokenizer/entities.test]: We don't report error in this case due to:

> If the character reference was consumed as part of an attribute (return state is either attribute value (double-quoted) state, attribute value (single-quoted) state or attribute value (unquoted) state), and the last character matched is not a U+003B SEMICOLON character (;), and the next input character is either a U+003D EQUALS SIGN character (=) or an ASCII alphanumeric, then, for historical reasons, switch to the character reference end state.

https://html.spec.whatwg.org/multipage/parsing.html#named-character-reference-state

Comment (zcorpan) [inline tokenizer/entities.test]: Hmmmmm. That should probably switch to the ambiguous ampersand state. https://html.spec.whatwg.org/#syntax-attribute-value says 

> Attribute values are a mixture of text and character references, except with the additional restriction that the text cannot contain an ambiguous ampersand.

Comment (zcorpan) [inline tokenizer/entities.test]: https://github.com/whatwg/html/issues/2776

**Comment by zcorpan** [comment]
Rebased and fixed the `&noti;` in attribute case to not be a parse error.

File: tree-construction/entities01.dat [hunk 2]
```diff
@@ -68,6 +68,32 @@ I'm &notin; I tell you
 |     "I'm ∉ I tell you"
 
 #data
+&ammmp;
+#errors
+(1,1): expected-doctype-but-got-chars
+(1,7): unknown-named-character-reference
+#new-errors
+(1:7) missing-semicolon-after-character-reference
```
Comment (inikulin) [inline tree-construction/entities01.dat]: It should be `unknown-named-character-reference`

File: tree-construction/entities01.dat [hunk 3]
```diff
@@ -68,6 +68,32 @@ I'm &notin; I tell you
...
+|   <body>
+|     "&ammmp;"
+
+#data
+&ammmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmp;
+#errors
+(1,1): expected-doctype-but-got-chars
+(1,950): unknown-named-character-reference
+#new-errors
+(1:950) missing-semicolon-after-character-reference
```
Comment (inikulin) [inline tree-construction/entities01.dat]: Same here

**Comment by zcorpan** [comment]
Fixed.

**Pull request merged**